File: frontend/src/lib/grokPlan.test.mjs
import assert from "node:assert/strict";
import test from "node:test";

import {
  grokPlanFilterCategory,
  resolveAccountGrokPlan,
  resolveGrokPlan,
} from "./grokPlan.ts";

test("Grok tier mapping preserves key, display, paid, and billing metadata", () => {
  const cases = [
    [0, { key: "free", display: "Free", paid: false, billing: false }],
    [1, { key: "supergrok", display: "SuperGrok", paid: true, billing: true }],
    [2, { key: "x_basic", display: "X Basic", paid: true, billing: false }],
    [3, { key: "x_premium", display: "X Premium", paid: true, billing: true }],
    [4, { key: "x_premium_plus", display: "X Premium+", paid: true, billing: true }],
    [5, { key: "supergrok_heavy", display: "SuperGrok Heavy", paid: true, billing: true }],
    [6, { key: "supergrok_lite", display: "SuperGrok Lite", paid: true, billing: true }],
  ];

  for (const [tier, expected] of cases) {
    assert.deepEqual(resolveGrokPlan(tier), expected);
  }
});

test("Grok plans normalize persisted keys, legacy displays, and numeric tiers", () => {
  assert.equal(resolveGrokPlan("SuperGrok Heavy")?.key, "supergrok_heavy");
  assert.equal(resolveGrokPlan("X Premium+")?.key, "x_premium_plus");
  assert.equal(resolveGrokPlan("x_premium_plus")?.display, "X Premium+");
  assert.equal(resolveGrokPlan("6")?.display, "SuperGrok Lite");

  assert.deepEqual(resolveGrokPlan(9), {
    key: "9",
    display: "9",
    paid: true,
    billing: true,
  });
  for (const invalid of [-1, "", "api", "unknown", "not-a-tier", null]) {
    assert.equal(resolveGrokPlan(invalid), null);
  }
});

test("Grok account plan prefers backend metadata and keeps unknowns filterable", () => {
  const account = {
    plan_type: "free",
    grok_plan: {
      key: "x_basic",
      display: "X Basic",
      paid: true,
      billing: false,
    },
  };
  assert.equal(resolveAccountGrokPlan(account)?.key, "x_basic");
  assert.equal(grokPlanFilterCategory(account), "x_basic");
  assert.equal(grokPlanFilterCategory({ plan_type: "9" }), "other");
  assert.equal(grokPlanFilterCategory({ plan_type: "api" }), "other");
});

File: auth/grok_plan_test.go
package auth

import (
	"encoding/json"
	"math"
	"testing"
)

func TestGrokPlanFromTier(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name   string
		tier   any
		want   GrokPlan
		wantOK bool
	}{
		{"free", 0, GrokPlan{Key: "free", Display: "Free"}, true},
		{"supergrok", 1, GrokPlan{Key: "supergrok", Display: "SuperGrok", Paid: true, Billing: true}, true},
		{"x basic", 2, GrokPlan{Key: "x_basic", Display: "X Basic", Paid: true}, true},
		{"x premium", 3, GrokPlan{Key: "x_premium", Display: "X Premium", Paid: true, Billing: true}, true},
		{"x premium plus", 4, GrokPlan{Key: "x_premium_plus", Display: "X Premium+", Paid: true, Billing: true}, true},
		{"supergrok heavy", 5, GrokPlan{Key: "supergrok_heavy", Display: "SuperGrok Heavy", Paid: true, Billing: true}, true},
		{"supergrok lite", 6, GrokPlan{Key: "supergrok_lite", Display: "SuperGrok Lite", Paid: true, Billing: true}, true},
		{"numeric string", "6", GrokPlan{Key: "supergrok_lite", Display: "SuperGrok Lite", Paid: true, Billing: true}, true},
		{"future positive tier", 7, GrokPlan{Key: "7", Display: "7", Paid: true, Billing: true}, true},
		{"future decimal tier", json.Number("7.5"), GrokPlan{Key: "7.5", Display: "7.5", Paid: true, Billing: true}, true},
		{"negative tier", -1, GrokPlan{}, false},
		{"missing tier", nil, GrokPlan{}, false},
		{"invalid tier", "invalid", GrokPlan{}, false},
		{"nan tier", math.NaN(), GrokPlan{}, false},
	}

	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()
			got, ok := GrokPlanFromTier(tt.tier)
			if ok != tt.wantOK {
				t.Fatalf("GrokPlanFromTier(%v) ok = %v, want %v", tt.tier, ok, tt.wantOK)
			}
			if got != tt.want {
				t.Fatalf("GrokPlanFromTier(%v) = %#v, want %#v", tt.tier, got, tt.want)
			}
		})
	}
}

func TestResolveGrokPlanAliases(t *testing.T) {
	t.Parallel()

	tests := map[string]string{
		"free":            "free",
		"SuperGrok":       "supergrok",
		"X Basic":         "x_basic",
		"x_premium":       "x_premium",
		"X Premium+":      "x_premium_plus",
		"X Premium Plus":  "x_premium_plus",
		"SuperGrok Heavy": "supergrok_heavy",
		"supergrok_lite":  "supergrok_lite",
		"6":               "supergrok_lite",
	}
	for input, wantKey := range tests {
		input, wantKey := input, wantKey
		t.Run(input, func(t *testing.T) {
			t.Parallel()
			got, ok := ResolveGrokPlan(input)
			if !ok {
				t.Fatalf("ResolveGrokPlan(%q) returned !ok", input)
			}
			if got.Key != wantKey {
				t.Fatalf("ResolveGrokPlan(%q).Key = %q, want %q", input, got.Key, wantKey)
			}
		})
	}

	for _, input := range []any{"api", "unknown", "", "-2", nil} {
		if got, ok := ResolveGrokPlan(input); ok {
			t.Fatalf("ResolveGrokPlan(%v) = %#v, want invalid", input, got)
		}
	}
}

func TestGrokPlanFromAccessToken(t *testing.T) {
	t.Parallel()

	token := makeJWT(map[string]any{"sub": "user-1", "tier": 5})
	plan, ok := GrokPlanFromAccessToken(token)
	if !ok {
		t.Fatal("GrokPlanFromAccessToken returned !ok")
	}
	if plan.Key != "supergrok_heavy" || plan.Display != "SuperGrok Heavy" || !plan.Paid || !plan.Billing {
		t.Fatalf("unexpected plan: %#v", plan)
	}
	if got := GrokPlanTypeFromAccessToken("not-a-jwt"); got != "" {
		t.Fatalf("invalid token plan = %q, want empty", got)
	}
}

func TestParseGrokAuthJSONExtractsPlanType(t *testing.T) {
	t.Parallel()

	accessToken := makeJWT(map[string]any{"sub": "user-1", "tier": 4})
	raw, err := json.Marshal(map[string]any{
		"access_token":  accessToken,
		"refresh_token": "rt-test",
		"client_id":     "client-test",
	})
	if err != nil {
		t.Fatal(err)
	}
	credentials, err := ParseGrokAuthJSON(raw)
	if err != nil {
		t.Fatalf("ParseGrokAuthJSON: %v", err)
	}
	if len(credentials) != 1 || credentials[0].PlanType != "x_premium_plus" {
		t.Fatalf("parsed credentials = %#v, want x_premium_plus", credentials)
	}
}
